### ISSUE TITLE
Job ID usage

### ISSUE THREAD

**Issue by jameshcorbett** [ExaWorks/job-api-spec#115]
The specification states that 

> The ID [of a job] is assigned automatically by the implementation when the Job object is constructed. The ID is guaranteed to be unique on the client machine. The ID does not have to match the ID of the underlying LRM job, but is used to identify Job instances as seen by a client application.

What is the use-case here? The ID "identifies Job instances as seen by a client application" but isn't this something users could do themselves, by associating an integer or a UUID with each Job? 

Also, the ID _can't_ feasibly match the ID of the underlying LRM job because the LRM ID isn't available until after the `Job` is constructed.

**Comment by hategan** [comment]
> What is the use-case here? The ID "identifies Job instances as seen by a client application" but isn't this something users could do themselves, by associating an integer or a UUID with each Job?

Such an ID would not be available for any logging messages coming from within JPSI though. Which would make tracking jobs in logs pretty difficult if not impossible.

> Also, the ID _can't_ feasibly match the ID of the underlying LRM job because the LRM ID isn't available until after the `Job` is constructed.

Right. Hence the "does not have to match". It's written from the perspective of the user of the API rather than the implementer. But yes, it can't match the native ID given that it has to exist when the job is constructed.

**Comment by jameshcorbett** [comment]
> Hence the "does not have to match".

I think something like "the ID is in no way connected with the ID of the underlying LRM job" would be more clear. Right now, the way it's phrased, it sounds like it might be up to the implementer to decide whether to match the `Job` ID to the LRM ID. But it's not.

> Such an ID would not be available for any logging messages coming from within JPSI though. Which would make tracking jobs in logs pretty difficult if not impossible.

If that's the use-case, though, the constraint could be reasonably relaxed so that the ID is only unique process-wide. And perhaps then relaxed further so that the ID is only guaranteed to be unique at any one particular point in time. But then in Python (for instance) why not just use `id()` and `repr()`, which is the standard thing to do?

But even aside from language-specific reasons, I think this ID stuff should be up to the implementation. If there isn't a well-defined and necessary reason for it, why keep it in there? It could always be added later if it was found to be necessary. Easier to add than remove.

**Comment by hategan** [comment]
> [...] be up to the implementer to decide whether to match the `Job` ID to the LRM ID. But it's not.

That's a pretty strong statement that I'm not willing to make. For a traditional LRM, perhaps. In the most general case, I see no reason to exclude the possibility that a unique ID be pushed to a native ID.

> If that's the use-case, though, the constraint could be reasonably relaxed so that the ID is only unique process-wide.

I don't know about that. In the days of the grid it was pretty common to have a command line tool to submit a job such that you could quickly hack a "workflow" through a shell script. In such a scenario, each job would be created in a separate process, so you might end up with all the jobs having the same ID.

I think unique IDs on one machine are fairly easy to generate and the possibility of JPSI being used in complex ways would suggest that we should err on the side of caution.

Anyway, those are my thoughts; @andre-merzky @SteVwonder?

**Comment by jameshcorbett** [comment]
> That's a pretty strong statement that I'm not willing to make. For a traditional LRM, perhaps. In the most general case, I see no reason to exclude the possibility that a unique ID be pushed to a native ID.

That is a good point, I hadn't thought of the potential for LRMs to accept an external job ID.

> In the days of the grid it was pretty common to have a command line tool to submit a job such that you could quickly hack a "workflow" through a shell script. In such a scenario, each job would be created in a separate process, so you might end up with all the jobs having the same ID.

That is a use-case where users could easily add that feature, though.

**Comment by SteVwonder** [comment]
> Anyway, those are my thoughts; @andre-merzky @SteVwonder?

I don't think the following really helps or clarifies anything.  It just represents my thoughts on the topic:

The J/PSI ID doesn't sit too well with me either, only because it is a whole new thing that J/PSI is introducing without a clean/intuitive connection to the underlying IDs.  For the simple/common case, the J/PSI ID does help by abstracting away the need to worry about guaranteeing uniqueness across connectors.  For more complex cases, like if a user wants to reconnect to a job after restarting their workflow manager, they will need to store the underlying job IDs (and the connector/executor they came from).  So the abstraction doesn't help in that use case, and really only adds complexity as far as I can tell.

I don't have a great alternative to suggest.  The only thing that I can think of would be a combination of connector, executor, and job information (e.g., `Slurm[quartz.llnl.gov]-srun-16438.7`).  For logging purposes, a single ID is nicer than that implementation-dependent accumulation of info, but the single ID does make following the chain of connections harder in the case that you want to manually inspect/debug things.  It is also much easier to standardize a single ID as opposed to an agglomeration of various values (whose contents may vary between connectors and executors).

**Comment by hategan** [comment]
> The J/PSI ID doesn't sit too well with me either, only because it is a whole new thing that J/PSI is introducing without a clean/intuitive connection to the underlying IDs.

So does PBS and pretty much all LRMs. They introduce job ids without a very good reason except for the user to keep track and interact with the job using a unique handle.

**Comment by jameshcorbett** [comment]
My complaint about the J/PSI ID is just that it is a weird outgrowth, as the spec now stands. It doesn't have any defined usage within the interface (I mean, there are no methods that take a J/PSI ID as an argument) and it doesn't make a lot of sense to me _outside_ the interface either. Certainly unique IDs are useful for logging, but the interface doesn't specify how logging is to be done. That seems like a fine thing to leave up to the implementation, anyway. And some languages (like Python) have their own way of handling logging and debugging.

If methods were added that needed a J/PSI ID, then that would be a great reason to have it. But it sounds like the `attach` method under discussion (#81) is going to use the native ID. So until more use-cases for the J/PSI ID are found, I just don't like it.

**Comment by hategan** [comment]
> My complaint about the J/PSI ID is just that it is a weird outgrowth,

Sorry James. IDs are a very common tool when one needs to distinguish between objects that would otherwise be virtually indistinguishable. There is no requirement that the API have methods that have such IDs as parameters for the above role to be fulfilled. Just like there is no biological use for names, but they are terribly useful in society.

As for use cases, besides what I mentioned, virtually every higher level tool that I know of has some form of IDs for it's jobs. This includes Swift and PARSL. And virtually every library like PSI/J has job IDs. This includes Swift/CoG, SAGA, DRMAA, gLite, etc. IDs are so pervasive that I'm really scratching my head over this conversation.

**Comment by jameshcorbett** [comment]
Sorry, I'm still not seeing it. You mention distinguishing objects "that would otherwise be virtually identical." If I'm a programmer and I want to distinguish my `Job` objects, why should I write `job_one.get_id() == job_two.get_id()` when I could just write `job_one == job_two`? (And I'm not thinking of overwriting the equality operator.) I really don't think the naming parallel is a good one, because the "name" of any object is already in this way built in to the language.

As for use-cases, Yeah, Flux has job IDs, but they actually have a very specific use. You pass the ID through to various functions to wait on a job or to perform other operations. Slurm also has job IDs. But again, you use the job ID on the command line to manipulate your job. This is what is missing here. You mentioned earlier that LRMs

> introduce job ids without a very good reason except for the user to keep track and interact with the job using a unique handle.

But that seems like a _very_ good reason to me.

Python's `concurrent.futures` package is a kind of higher-level tool that doesn't have job IDs because there is no need for them. I feel like the job ID in J/PSI is like proposing that `concurrent.futures` add an ID to it's `Future` class.

**Comment by hategan** [comment]
> when I could just write `job_one == job_two`

In languages that have pointers and where "==" means pointer equality. I'm not sure I want to restrict the spec to just those. You seem to want to implement IDs implicitly using language features. That's fine-ish. It suffers from some problems, such possible collisions, which is why we opted for host uniqueness, although I'm inclined to suggest global uniqueness.

> I really don't think the naming parallel is a good one, because the "name" of any object is already in this way built in to the language.

We're not aiming for "the language", but for multiple possible languages.

> As for use-cases, Yeah, Flux has job IDs, but they actually have a very specific use. You pass the ID through to various functions to wait on a job or to perform other operations. Slurm also has job IDs. But again, you use the job ID on the command line to manipulate your job. This is what is missing here.

That's fine. As I said, it's useful beyond that. This is a library that will be used by other systems. There will also be Layers 1 and 2, which are distributed systems, which will have services to manipulate jobs, and IDs will be essential there.

> The difference with J/PSI is that the Job instance is the unique handle.

Besides the issue above with this being language specific, log files. The Job instance has no textual meaning. It must be converted to textual form to appear in a log. That's where you see the ID. And then you can grep through that log for that ID and see what the job has done, from inception to wherever it goes.

> Python's concurrent.futures package is a kind of higher-level tool that doesn't have job IDs 

Sorry I wasn't clear. Higher level than PSI/J. Something that uses something like PSI/J. Something that we're hoping will adopt PSI/J. The examples I gave would, in order to adopt PSI/J, probably need something like job IDs so will either use if available or need to emulate them.

> so if you could explain to me why J/PSI is not analogous to concurrent.futures that would be very helpful.

I gave you examples of libraries with the almost exact same purpose as PSI/J. I am not sure why something that's much farther from it would be more relevant.

Let's try something different: propose some clear changes and go with the idea that the burden is on you to demonstrate why the proposed changes should be adopted. If presented with challenges, try to address those challenges specifically, without introducing lateral issues into the discussion. We can then discuss and put it to a vote if it's not quite clear where it should go.

And just to be clear here: I don't think the arguments for having the ID are terribly compelling. They are not. It's just that the arguments against it seem even less compelling.

**Comment by jameshcorbett** [comment]
> We're not aiming for "the language", but for multiple possible languages.

Right, but different languages can have additions to the spec. If language X doesn't have some way of testing object identity and need to compare ID strings instead, then language X should have these job ID strings. Other languages don't need them. You make it sound like the spec has to be broad enough to cover every feature that every programming language might need. If you were serious about that, you would get a very cluttered interface...

> distributed systems, which will have services to manipulate jobs, and IDs will be essential there.

Right, but my point all along is that the spec _as it stands now_ does not allow you to do anything with the job ID. How are you supposed to manipulate jobs with the ID? It's not like you can fetch a job given an ID or anything like that. (But with `attach` you can manipulate a job with the _native_ ID---I don't have any problem with the native ID and I never have.)

> I gave you examples of libraries with the almost exact same purpose as PSI/J. I am not sure why something that's much farther from it would be more relevant.

You didn't understand. Just throwing the names out of a bunch of libraries is not helpful. I want to know _why_ the job ID is a good fit for those libraries. I imagine they all have well-defined use-cases where you can plug the ID in somewhere and get something out somewhere else.

> log files. The Job instance has no textual meaning. It must be converted to textual form to appear in a log.

Again, this is something that languages might best do in very different ways. In Python I would think that the best thing would be something like `<Job Slurm[quartz.llnl.gov]-srun-16438.7 at 0xff435a4 reached state ACTIVE>` would be the most helpful. And this you could with `__repr__` or `__str__`.

> Let's try something different: propose some clear changes and go with the idea that the burden is on you to demonstrate why the proposed changes should be adopted.

#120

**Comment by hategan** [comment]
>Just throwing the names out of a bunch of libraries is not helpful.

That's inappropriate. Please don't do it again.

> I want to know why the job ID is a good fit for those libraries

It is not my responsibility to teach you that. Please go and study those libraries, their respective papers, talk to their authors, etc.

In the mean time, I am closing this issue without further comment.

**Comment by andre-merzky** [comment]
I tend to agree with @jameshcorbett on this question.  I am not sure if that means that JobIDs should be removed, but I am unclear about their use.  As James stated: different languages have different means to identify object instances, so we don't need to specify a JobID (different than the native backend ID) for that purpose.  But the JobID is also not used in any other part of the API, so why do we have it?

The libraries listed by @hategan indeed all do have JobIDs in their API, but in all those cases (AFAICS [*]), that ID is in fact used to identify the backend job instance, not the API object instance: any Job object re-attaching to the same backend job would have the same JobID.  J/PSI uses the `NativeJobID` for that purpose.

> In the mean time, I am closing this issue without further comment.

With all due respect, I personally don't think we should close issues like this.

[*]  I am actually not sure for PARSL - can it re-attach to jobs/tasks?  I am positive for CoG, DRMAA, SAGA, Radical-Pilot, gLite, Globus, POSIX (and any API I know which build over POSIX, like the Python process APIs etc).

**Comment by SteVwonder** [comment]
> As James stated: different languages have different means to identify object instances, so we don't need to specify a JobID (different than the native backend ID) for that purpose. But the JobID is also not used in any other part of the API, so why do we have it?

I agree with this reasoning: if you aren't using it, why have it?  That being said, maybe we have some things on the horizon that will require it.  While looking into one of the workflow systems listed above (I think gLite), I realized that some of those systems have CLIs that consume the workflow system's ID as opposed to the native ID. I don't want to divert this issue into a "will we have a CLI or not" discussion, but *if* we do add a J/PSI CLI, we will almost certainly need a J/PSI ID.

Re-reading the discussion, I see @hategan pointing out a similar thought:
> There will also be Layers 1 and 2, which are distributed systems, which will have services to manipulate jobs, and IDs will be essential there.

I'm entirely inexperienced on systems that operate at Layer 1 (and only familiar with Layer 2 in-so-far as Flux can be considered a nested scheduler).  So I am not really able to expand on the services that would require an ID in Layers 1 & 2.  @hategan: maybe you can elaborate on one use case for J/PSI IDs at Layer 1 or 2?  Personally, that would convince me that we leave things as they are (i.e., keep J/PSI IDs in the API) since we will need them for the next version anyways.

**Comment by hategan** [comment]
> I agree with this reasoning: if you aren't using it, why have it?

That is an incorrect assumption. The API doesn't use it. Like the API itself, it is meant to be used by higher-level components. So I think there is some confusion happening here between internal and external use. By definition, the whole purpose of the API is external use and "internal" is only there in the sense that not all entities in the API are entirely isolated. But to state that an isolated entity in an API is grounds for exclusion of that entity is silly: you may very well exclude the entire API by carefully adjusting the definition of "entity". 

In short, "aren't using it" is incorrect. The ID is used to uniquely identify jobs in a distributed/multi-component system.

> That being said, maybe we have some things on the horizon that will require it.

No, we don't have anything that will "require" it. You are setting the bar artificially high and, if you carefully go through the discussion, you will see that "required" was never the case. The point of a universal identifier in a multi-component and distributed system is convenience. So if you've worked with such systems, you may even argue that it's terribly useful convenience. You could generally have identifiers that are only shared by individual connections between components, but you'd have to maintain separate identifiers for each such connection, as well as a mechanism for associating them with the actual objects. 

> So I am not really able to expand on the services that would require an ID in Layers 1 & 2.

Resource identifiers for REST resources, perhaps?

>  Personally, that would convince me that we leave things as they are

I believe that to be the wrong approach, as I pointed out to @jameshcorbett. If you propose a change, the burden should be on you to justify it. It sounds a bit cheeky to me to demand a change without some valid justification.

**Comment by hategan** [comment]
> > In the mean time, I am closing this issue without further comment.
> 
> With all due respect, I personally don't think we should close issues like this.

You are forcing me to spell it out.

I closed the issue because of "Just throwing the names out of a bunch of libraries is not helpful.[...] I imagine they all have well-defined use-cases". Those are not some random libraries, but the rather few libraries that have attempted to do what we are doing with this API. So they are THE libraries that we should look at. I find it a bit disrespectful to 1) not bother to do a modicum of research about those libraries given that we've been talking about them for 6+ months now 2) not having done (1), suggest that your interlocutor must not know what they are talking about, and, 3), substitute imagination for research to justify (2).

Discussions can get heated and, in that heat, such things can happen more or less willingly. That's fine. But if they go way beyond not useful and into wasteful territory, it's best if we stop pursuing them.

**Comment by andre-merzky** [comment]
> > > In the mean time, I am closing this issue without further comment.
> > 
> > 
> > With all due respect, I personally don't think we should close issues like this.
> 
> You are forcing me to spell it out.

I do, and thanks for playing along.  I understand that you are frustrated, so double-thanks!

> I closed the issue because of "Just throwing the names out of a bunch of libraries is not helpful.[...] I imagine they all have well-defined use-cases". Those are not some random libraries, but the rather few libraries that have attempted to do what we are doing with this API. So they are THE libraries that we should look at. I find it a bit disrespectful to 1) not bother to do a modicum of research about those libraries given that we've been talking about them for 6+ months now

I am fairly sure I did look at those libraries (I worked with all of them, and contributed to specification and code for 3 of them), and still come to the same point James is (I think) trying to make.  This discussion might not be very exciting, and ultimately may not lead to any changes - but due diligence is being followed...

**Comment by andre-merzky** [comment]
> In short, "aren't using it" is incorrect. 

You make a fair point: most parts of the API are not used by the API itself but still considered useful.

> The ID is used to uniquely identify jobs in a distributed/multi-component system.

They don't though - they only uniquely identify the object instance.

```py
je = jpsi.Executor(...)
j1 = jpsi.Job(...)
je,submit(j1)
id_1 = j1.get_id()
nid_1 = j1.get_native_id()

j2 = jpsi.Job()
je.attach(j2, nid_1)
id_2 = j2.get_id()
nid_2 = j2.get_native_id()

assert(nid_1 == nid_2)  # True
assert(id_1 == id_2)  # False
```

How would you use the jpsi ID?  When using them to identify jobs, the above would now have two IDs identifying the same job - which is not very intuitive.  If I would use the API for managing many jobs from different backends, I would most likely identify them by the tuple `[je.url, j.native_id]` - that information is unique per job, and can be used to reconnect.  The jpsi ID does neither.  In that sense, it is about as useful as `assert(id(j1) != id(j2))` - which is to say that languages already provide that.

That is different to SAGA, Globus, gLite, CoG, DRMAA and friends - there the ID does indeed uniquely identify the job, not the object instance.

**Comment by hategan** [comment]
> I am fairly sure I did look at those libraries

Didn't claim you didn't.

> but in all those cases (AFAICS [*]), that ID is in fact used to identify the backend job instance, not the API object instance

Some of these APIs use the native ID directly.
Some, like PSI-J, have a Job ID which, through some mapping, can be used to identify backend jobs.
It's not clear to me what the statement is here.

> [*] I am actually not sure for PARSL - can it re-attach to jobs/tasks? I am positive for CoG, DRMAA, SAGA, Radical-Pilot, gLite, Globus, POSIX (and any API I know which build over POSIX, like the Python process APIs etc).

I'm not sure whether re-attaching is relevant. Perhaps the question we want to ask is: If we lose the Job ID, can we still interact with the job through the API? This isn't true for PARSL, CoG, Globus >= 3, and others. But this is also very weak. One can persist the Job ID - Native ID mapping to change the answer and many of these libraries may internally have such provisions already (e.g., they keep jobs in a DB, etc.).

**Comment by hategan** [comment]
> They don't though - they only uniquely identify the object instance.

That's a valid problem. Attaching makes this a many-to-one mapping rather than 1-to-1. It was added as an after-though and the current spec is silent about what should happen to the ID when you re-attach.

> I would most likely identify them by the tuple [je.url, j.native_id]

We need to stop re-litigating issues if we are to make progress. Please look at the very first message in this thread. The native_id only becomes available after submission to a native backend. The url only becomes available after submission to a remote service. Any client-side queuing that happens before would leave the job without an ID. You can't use let-binding IDs as universal identifiers for a job since they are not even guaranteed to happen.

> That is different to SAGA, Globus, gLite, CoG, DRMAA and friends - there the ID does indeed uniquely identify the job, not the object instance.

Not true for CoG. Mixed for Globus >= 3 to the extent that you can't re-attach to a native ID, but the Job ID is generated on the client side. So if it supported re-attachment, this wouldn't hold. I'm surprised you listed RP earlier. The CU ID is also generated at CU initialization.

> That is exactly the case where things would break: a service restart would result in different job IDs, even though the native IDs and the underlying jobs remain valid.

Not really sure what you are saying here. If designed to be re-startable, the service will need to persist the mapping between resource IDs and backend IDs. Assuming, while the service is alive, that jobs states are kept in PSI-J objects, when re-loading objects after a restart, the whole Job, including the ID, would be de-serialized. Although we'll have to see. If using an existing REST framework, jobs will likely be kept in a DB to begin with and resource objects only created during REST API calls if at all.

**Comment by andre-merzky** [comment]
> Not true for CoG. Mixed for Globus >= 3 to the extent that you can't re-attach to a native ID, but the Job ID is generated on the client side. So if it supported re-attachment, this wouldn't hold. I'm surprised you listed RP earlier. The CU ID is also generated at CU initialization.

The version of CoG I know is fairly old - but it just used the globus job ID.  In `src/cog/modules/provider-gt2ft/src/org/globus/cog/abstraction/impl/execution/gt2ft/JobSubmissionTaskHandler.java`:

```java
this.gramJob.setID((String) task.getAttribute("globusID"));
```

That ID is basically formed like `<gatekeeper_url>/<gatekeeper_jobid>`, so the ID was, from the perspective of  CoG, a unique backend ID.  That's what I had in mind at least.  This may have changed later...

> I'm surprised you listed RP earlier. The CU ID is also generated at CU initialization.

>> The RP task ID does persist across reconnect, you can retrieve a task object instance via that ID, and that object will then represent the same user task. It is not a question of where the ID is created, but rather what it identifies...

> We need to stop re-litigating issues if we are to make progress. Please look at the very first message in this thread. The native_id only becomes available after submission to a native backend.

Apologies, I do not seem to get my message across.  Let me try again.

I do not want to argue that the native ID can be used to identify object instances - it can not.  You are right, it is not even available until several state transitions.

I argue that the jpsi job ID is not needed, as language native means are sufficient.  Specifically in Python: `id(job)` provides the exact same functionality.  This leads me to question the need for the jpsi job ID.   I do not see any use case where it would be useful to have.

**Comment by hategan** [comment]
> gt2ft/JobSubmissionTaskHandler.java

That was an experimental "gt2 - fault tolerant" provider. This isn't done in the GT2 provider or any others that I know. And it's an implementation choice, not specified anywhere.

> I argue that the jpsi job ID is not needed, as language native means are sufficient. Specifically in Python: `id(job)` provides the exact same functionality. This leads me to question the need for the jpsi job ID. I do not see any use case where it would be useful to have.

I think we are conflating not having IDs with using the platform to implement the ID generation.

Using the platform is a choice. But, again, please go back to the discussion for why this isn't such a good idea:
- it's not even guaranteed to be unique in the same process; only in the same process at one instance in time
- you can't move a job from a client to a server and expect it to have anywhere near the same ID if the two are implemented in different languages

**Comment by hategan** [comment]
I seriously have no more time to say the same things over and over. Please, be considerate.

**Comment by andre-merzky** [comment]
> That was an experimental "gt2 - fault tolerant" provider...

Ha, funny that this is what stuck in my brain :-D

> it's not even guaranteed to be unique in the same process; only in the same process at one instance in time

True - they are unique during the lifetime of the object.  That's the same the spec promises for jpsi IDs though?

> you can't move a job from a client to a server and expect it to have anywhere near the same ID if the two are implemented in different languages

I don't understand this point - how does that relate to jpsi IDs?  We do not specify any serialization etc, so we can't move objects either  way?  Do we have that use case?

> I seriously have no more time to say the same things over and over. Please, be considerate.

ok then.

**Comment by hategan** [comment]
> > it's not even guaranteed to be unique in the same process; only in the same process at one instance in time
> 
> True - they are unique during the lifetime of the object. That's the same the spec promises for jpsi IDs though?

The spec promises machine-unique. I think we should consider globally unique.

> > you can't move a job from a client to a server and expect it to have anywhere near the same ID if the two are implemented in different languages
> 
> I don't understand this point - how does that relate to jpsi IDs? We do not specify any serialization etc, so we can't move objects either way? Do we have that use case?

We're going to have to move objects in the other layers, no?

And, again, this is something that's meant to be used in scalable and distributed systems. Having objects moved around is expected and should be designed for.
